feat(http): Add X-Permitted-Cross-Domain-Policies page (#37192)
mdn/content: files/en-us/web/http/headers/x-permitted-cross-domain-policies/index.md

---
title: X-Permitted-Cross-Domain-Policies
slug: Web/HTTP/Headers/X-Permitted-Cross-Domain-Policies
page-type: http-header
status:
  - non-standard
---

{{HTTPSidebar}}

> The HTTP **`X-Permitted-Cross-Domain-Policies`** {{Glossary("response header")}} specifies if a cross-domain policy file (`crossdomain.xml`) is allowed.
> 
> A cross-domain policy file may grant clients, such as Adobe Acrobat or Apache Flex, permission to handle data across domains that would otherwise be restricted due to the [Same-Origin Policy](/en-US/docs/Web/Security/Same-origin_policy).
> The `X-Permitted-Cross-Domain-Policies` header overrides such policy files so that clients still block unwanted requests.
> 
> Usage of this header is less common since Adobe Flash Player and Microsoft Silverlight have been deprecated.

Reviewer: I'm not sure this is helpful or correct - specifically, this allows you override the policy files, if any, that are used, and then these policy files grant permissions to access the site. I'd also push some of this into the description.

```suggestion
The HTTP **`X-Permitted-Cross-Domain-Policies`** {{Glossary("response header")}} enables a server to override the meta-policy for a site, indirectly controlling whether the site can be accessed cross-origin by a document running in a web client like Adobe Acrobat or Microsoft Silverlight.

Usage of this header is less common since Adobe Flash Player and Microsoft Silverlight have been deprecated.
```

See below for desc changes.

Author: brilliant, thank you

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>{{Glossary("Response header")}}</td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden header name")}}</th>
      <td>No</td>
    </tr>
  </tbody>
</table>

## Syntax

```http

> X-Permitted-Cross-Domain-Policies: <permitted-cross-domain-policies>

Reviewer: As I read the spec, this header is an alternative way to set the meta policy, and only _one_ meta policy can be set. Please check.

```suggestion
X-Permitted-Cross-Domain-Policies: <permitted-cross-domain-policy>
```

Author: Yes, that's correct!

```

## Directives

- `none`
  - : No policy files are allowed anywhere on the target server, including in a master policy file.
- `master-only`
  - : Allows cross-domain access to the master policy file defined on the same domain.
- `by-content-type` (HTTP/HTTPS only)
  - : Only policy files served with `Content-Type: text/x-cross-domain-policy` are allowed.
- `by-ftp-filename` (FTP only)
  - : Only policy files whose file names are `crossdomain.xml` (URLs ending in `/crossdomain.xml`) are allowed.
- `all`
  - : All policy files on this target domain are allowed.
- `none-this-response`
  - : Indicates that the current document should not be used as a policy file despite other headers or its content.
    This value is unique to the HTTP header only.

## Description

> Policy files are defined in a `crossdomain.xml` file, typically located at the root of the domain (`http://example.com/crossdomain.xml`) and define policies as so (the least permissive master policy definition):
> 
> ```xml
> <?xml version="1.0"?>
> <!DOCTYPE cross-domain-policy SYSTEM "http://www.adobe.com/xml/dtds/cross-domain-policy.dtd">
> <cross-domain-policy>
>   <site-control permitted-cross-domain-policies="none"/>
> </cross-domain-policy>
> ```
> 
> Instead of relying on policy files, clients may check for the `X-Permitted-Cross-Domain-Policies` header to detect a meta-policy.
> The purpose of the HTTP header is for cases where people wish to declare a cross-domain policy, but cannot write to the root directory of the domain.

Reviewer: crossdomain.xml is the master policy file, which defines what other ones are allowed, if any. It is this file the header overrides.

````suggestion
Web clients such as Adobe Acrobat or Apache Flex can load web documents, which may in turn load resources from the same site or other sites.
Access is restricted to same-site resources by default, due to the [same origin policy](/en-US/docs/Web/Security/Same-origin_policy), but cross-origin sites may choose to make some or all of their resources available to clients cross-origin using special files, referred to as a cross-domain policy files.

A "master" cross-domain policy file may be defined as `crossdomain.xml` file in the root of the domain, for example: `http://example.com/crossdomain.xml`.
The master file defines the _meta-policy_ for the whole site using the `permitted-cross-domain-policies` attribute of the  `site-control` tag.
The meta-policy controls whether any policies are allowed, and the conditions for the the other "sub" cross-domain policy files to be used. These other policy files might be created in particular directories to specify access over the files in their given directory tree.

For example, this is the least permissive master policy definition, which does not permit any access, and does not allow the use of other "sub" policy files.

```xml
<?xml version="1.0"?>
<!DOCTYPE cross-domain-policy SYSTEM "http://www.adobe.com/xml/dtds/cross-domain-policy.dtd">
<cross-domain-policy>
  <site-control permitted-cross-domain-policies="none"/>
</cross-domain-policy>
```

The HTTP **`X-Permitted-Cross-Domain-Policies`** header can be used to override the master cross-domain policy file `permitted-cross-domain-policies` information.
It can be used in cases where the website needs to declare a cross-domain policy, but cannot write to the root directory of the domain.
````

## Examples

### Disallowing cross-domain policy files

If you don't need to load application data in clients such as Adobe Flash Player or Adobe Acrobat (or legacy clients), then the header should be configured as `X-Permitted-Cross-Domain-Policies: none`:

```http
X-Permitted-Cross-Domain-Policies: none
```

> Some security testing tools will check for the presence of a `X-Permitted-Cross-Domain-Policies: none` header because it can mitigate the risk of an unwanted policy file added to your site by accident or through malicious actions.

Reviewer: Move into intro

```suggestion
```

A "master" cross-domain policy file may be defined as `crossdomain.xml` file in the root of the domain, for example: `http://example.com/crossdomain.xml`.
The master file defines the _meta-policy_ for the whole site using the `permitted-cross-domain-policies` attribute of the `<site-control>` tag.
The meta-policy controls whether any policies are allowed, and the conditions for the other "sub" cross-domain policy files to be used.
These other policy files might be created in particular directories to specify access over the files in their given directory tree.

For example, this is the least permissive master policy definition, which does not permit any access, and does not allow the use of other "sub" policy files.

```xml
<?xml version="1.0"?>
<!DOCTYPE cross-domain-policy SYSTEM "http://www.adobe.com/xml/dtds/cross-domain-policy.dtd">
<cross-domain-policy>
  <site-control permitted-cross-domain-policies="none"/>
</cross-domain-policy>
```

> The `X-Permitted-Cross-Domain-Policies` header can specify a meta-policy for the HTTP response it's included in or override a meta-policy defined in the master cross-domain policy file, if present.

Reviewer: ```suggestion
The `X-Permitted-Cross-Domain-Policies` header can specify a meta-policy for the HTTP response it's included in, or override a meta-policy defined in the master cross-domain policy file, if present.
```
